File: core/src/com/bombbird/terminalcontrol/entities/trafficmanager/DayNightManager.java
package com.bombbird.terminalcontrol.entities.trafficmanager;

import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.screens.RadarScreen;

import java.util.Calendar;
import java.util.TimeZone;

public class DayNightManager {
    public static final Array<String> NIGHT_AVAILABLE = new Array<>();

    private static void loadArray() {
        NIGHT_AVAILABLE.add("TCTT", "TCHH", "TCBB", "TCMD");
    }

    public static boolean checkNoiseAllowed(boolean night) {
        RadarScreen radarScreen = TerminalControl.radarScreen;
        if (!radarScreen.allowNight) return !night;
        Calendar calendar = Calendar.getInstance(TimeZone.getDefault());
        int additional = calendar.get(Calendar.AM_PM) == Calendar.PM ? 12 : 0;
        int time = (calendar.get(Calendar.HOUR) + additional) * 100 + calendar.get(Calendar.MINUTE);
        if (radarScreen.nightEnd < radarScreen.nightStart) {
            //Cross midnight, additional step needed
            if (time >= radarScreen.nightStart || time < radarScreen.nightEnd) {
                return night; //True if is night and star is night
            } else {
                return !night; //True if is day and star is day
            }
        } else {
            if (time >= radarScreen.nightStart && time < radarScreen.nightEnd) {
                return night; //True if is night and star is night
            } else {
                return !night; //True if is day and star is day
            }
        }
    }

    public static boolean isNightAvailable() {
        if (NIGHT_AVAILABLE.size == 0) loadArray();
        if (TerminalControl.radarScreen == null) return false;
        return NIGHT_AVAILABLE.contains(TerminalControl.radarScreen.mainName, false);
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/trafficmanager/DayNightManager.kt
package com.bombbird.terminalcontrol.entities.trafficmanager

import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.sidstar.SidStar
import java.util.*

object DayNightManager {
    private val NIGHT_AVAILABLE = Array<String>()

    private fun loadArray() {
        NIGHT_AVAILABLE.add("TCTT", "TCHH", "TCBB", "TCMD")
    }

    /** Checks if a SID/STAR is allowed depending on whether night mode is active */
    fun checkNoiseAllowed(sidStar: SidStar): Boolean {
        return if (isNight) sidStar.night else sidStar.day
    }

    /** Checks whether airport is utilising night operations */
    val isNight: Boolean
        get() {
            TerminalControl.radarScreen?.let {
                if (!isNightAvailable || !it.allowNight) return false
                val calendar = Calendar.getInstance(TimeZone.getDefault())
                val additional = if (calendar[Calendar.AM_PM] == Calendar.PM) 12 else 0
                val time = (calendar[Calendar.HOUR] + additional) * 100 + calendar[Calendar.MINUTE]
                return if (it.nightEnd <= it.nightStart) {
                    //Cross midnight
                    time >= it.nightStart || time < it.nightEnd
                } else {
                    time >= it.nightStart && time < it.nightEnd
                }
            }
            return false
        }

    /** Check whether this airport has night operations available */
    val isNightAvailable: Boolean
        get() {
            if (NIGHT_AVAILABLE.size == 0) loadArray()
            return TerminalControl.radarScreen?.let { NIGHT_AVAILABLE.contains(it.mainName, false) } ?: false
        }
}